### PR TITLE
feat(publishing): add Nexus publish plugin for publishing to maven central

### PR DESCRIPTION
**Pull request by danielpeach**

### CODE REVIEW

File: spinnaker-project-plugin/src/main/groovy/com/netflix/spinnaker/gradle/publishing/nexus/NexusPublishPlugin.groovy [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        nexusUrl = new URI(nexusExtension.nexusStagingUrl().get())
+        snapshotRepositoryUrl = new URI(nexusExtension.nexusSnapshotUrl().get())
+        username = nexusExtension.nexusUsername()
+        password = nexusExtension.nexusPassword()
+        stagingProfileId = nexusExtension.nexusStagingProfileId()
+      }
+    }
+  }
+
+  private void configureSpinnakerPublicationForNexus(Project project, NexusPublishExtension nexusExtension) {
```
Comment (danielpeach) [inline spinnaker-project-plugin/src/main/groovy/com/netflix/spinnaker/gradle/publishing/nexus/NexusPublishPlugin.groovy:76]: Configures the publication to match what Sonatype requires: javadoc jars, pgp signing for every artifact, and a pom with extra metadata.

File: spinnaker-project-plugin/src/main/groovy/com/netflix/spinnaker/gradle/publishing/nexus/NexusPublishPlugin.groovy [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+      def spinnakerPublication = publishingExtension.publications.getByName(PublishingPlugin.PUBLICATION_NAME)
+
+      configurePom(project, spinnakerPublication)
+
+      project.extensions.configure(SigningExtension) { signingExtension ->
+        signingExtension.useInMemoryPgpKeys(nexusExtension.pgpSigningKey(), nexusExtension.pgpSigningPassword())
+        signingExtension.sign(spinnakerPublication)
+      }
+
+      if (project.plugins.hasPlugin(DokkaPlugin)) {
```
Comment (danielpeach) [inline spinnaker-project-plugin/src/main/groovy/com/netflix/spinnaker/gradle/publishing/nexus/NexusPublishPlugin.groovy:87]: If we have dokka, use dokka. If not, fall back to javadoc. I gave up on groovy docs. A pure groovy subproject will generate an empty javadoc jar, but that's OK with Sonatype.

File: spinnaker-project-plugin/src/main/groovy/com/netflix/spinnaker/gradle/publishing/nexus/NexusPublishExtension.groovy [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  final Property<Boolean> enabled
+  final Property<String> nexusStagingUrl
+  final Property<String> nexusSnapshotUrl
+  final Property<String> nexusStagingProfileId
+
+  NexusPublishExtension(Project project) {
+    this.project = project
+    ObjectFactory props = project.objects
+    enabled = props.property(Boolean).convention(false)
+    nexusStagingUrl = props.property(String).convention("https://s01.oss.sonatype.org/service/local/")
```
Comment (danielpeach) [inline spinnaker-project-plugin/src/main/groovy/com/netflix/spinnaker/gradle/publishing/nexus/NexusPublishExtension.groovy:37]: This is the config (staging url, snapshot url, profile ID) we were assigned for the `io.spinnaker` group id.

**Review by danielpeach — COMMENTED** [review]
I tested that we pass the release requirements for Kork, Orca, and Fiat and tested a round-trip of a staged Kork inside Orca.

One thing I noticed is that we don't include maven central as a default repository. I can add in this PR or in a follow-up.

**Review by ajordens — APPROVED** [review]
LGTM -- asked @cfieber to also give it a look.

**Review by cfieber — APPROVED** [review]

**Pull request merged**